### PR TITLE
GH-36824: [C++] Improve the test tracing of CheckWithDifferentShapes in the if-else kernel tests

### LINKED ISSUE #36824: [C++] scalar_if_else_test.cc: The test traces in CheckWithDifferentShapes can be misleading
### Describe the enhancement requested

The traces might show that a call is being made with a conditional scalar, but then a call is made with an array. If there is a bug in the kernel, it's hard to know which specialization is to blame given the trace.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

The traces might show that a call is being made with a conditional scalar, but then a call is made with an array. If there is a bug in the kernel, it's hard to know which specialization is to blame given the trace.

### What changes are included in this PR?

A simplification of the `CheckWithDifferentShapes` and addition of a bit more tracing information.

### Are these changes tested?

Yes. The changes are only made to test code and I verified that bugs in the implementation get caught by the new code.

### Are there any user-facing changes?

No.
* Closes: #36824

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_if_else_test.cc [hunk 1]
```diff
@@ -126,75 +126,105 @@ TYPED_TEST(TestIfElsePrimitive, IfElseFixedSizeRand) {
...
+    }
+    return trace;
+  });
+  Datum cond_in;
+  Datum left_in;
+  Datum right_in;
+  int cond_index = 0;
+  int left_index = 0;
+  int right_index = 0;
+  while (NextScalarOrWholeArray(cond, &cond_index, &cond_in)) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:203]: `NextScalarOrWholeArray` returns false when it fills an array in its last argument... so the loop doesn't run in that case. I suppose that's not expected?

File: cpp/src/arrow/compute/kernels/scalar_if_else_test.cc [hunk 2]
```diff
@@ -126,75 +126,105 @@ TYPED_TEST(TestIfElsePrimitive, IfElseFixedSizeRand) {
...
-            ASSERT_OK_AND_ASSIGN(expected, IfElse(cond_bcast, left_bcast, right_bcast));
-          }
-          AssertDatumsEqual(expected, actual, /*verbose=*/true);
+    if (!left.type()->Equals(*right.type())) {
+      return Cast(expected, CastOptions::Safe(std::move(type)));
+    }
+    return expected;
+  }
+  // When at least one of the inputs is an array, we expect the output
+  // to be the same as if all the scalars were broadcast to arrays.
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:154]: If all inputs are array, then the expected datum will be trivially equal to the actual. Should we reuse the original `expected` argument in that case?

File: cpp/src/arrow/compute/kernels/scalar_if_else_test.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:203]: Good catch. An off-by-one bug. I will fix.

File: cpp/src/arrow/compute/kernels/scalar_if_else_test.cc (hunk 2 above)
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/scalar_if_else_test.cc:159]: Done.

**Review by pitrou — APPROVED** [review]
+1, let's just wait for CI

**Pull request merged**